samcli/local/lambdafn/runtime.py

```diff
@@ -36,9 +38,170 @@ def __init__(self, container_manager, image_builder):
             Instance of the ContainerManager class that can run a local Docker container
         image_builder samcli.local.docker.lambda_image.LambdaImage
             Instance of the LambdaImage class that can create am image
+        warm_containers bool
+            Determines if the warm containers is enabled or not.
         """
         self._container_manager = container_manager
         self._image_builder = image_builder
+        self._containers = {}
+        self._warm_containers = warm_containers
+        self._paths_to_be_cleaned = []
+
+        # used to observe any code change in case if warm containers is enabled to support the reloading
+        self._observed_paths = {}
+        self._observed_paths_checksum = {}
+        self._function_paths = {}
+        self._observer = Observer()
+        self._code_change_handler = PatternMatchingEventHandler(
+            patterns=["*"], ignore_patterns=["*.swp"], ignore_directories=False
+        )
+        self._code_change_handler.on_modified = self._on_code_change
+        self._lock = threading.Lock()
```

Comment:
Which shared resources are we guarding?

Reply:
the Observed code dir paths, and containers